### PR TITLE
Fixed #31765 -- Made SQLite connections disable legacy ALTER TABLE.

### PR DESCRIPTION
**Pull request by adamchainz**
Alternative fix for ticket-31765. Works on my macOS 11.6 machine.

### CODE REVIEW

File: django/db/backends/sqlite3/base.py [hunk 1]
```diff
@@ -259,6 +259,9 @@ def get_new_connection(self, conn_params):
         conn.create_aggregate('VAR_POP', 1, list_aggregate(statistics.pvariance))
         conn.create_aggregate('VAR_SAMP', 1, list_aggregate(statistics.variance))
         conn.execute('PRAGMA foreign_keys = ON')
+        # The macOS bundled SQLite defaults legacy_alter_table ON, which
+        # prevents atomic table renames (feature supports_atomic_references_rename)
+        conn.execute('PRAGMA legacy_alter_table = OFF')
```
Comment (carltongibson) [inline django/db/backends/sqlite3/base.py:264]: This was the _Right Fix™_ last time I looked into this issue. 

Typically I no longer see this locally... — I shall try to reproduce, but, provisionally, should be good.

File: django/db/backends/sqlite3/features.py [hunk 2]
```diff
@@ -91,10 +90,6 @@ def django_test_skips(self):
 
     @cached_property
     def supports_atomic_references_rename(self):
-        # SQLite 3.28.0 bundled with MacOS 10.15 does not support renaming
-        # references atomically.
-        if platform.mac_ver()[0].startswith('10.15.') and Database.sqlite_version_info == (3, 28, 0):
```
Comment (carltongibson) [inline django/db/backends/sqlite3/features.py:96]: Yes... our hope this was just the one version proved wrong.

Comment (adamchainz) [reply django/db/backends/sqlite3/features.py:96]: Ah whilst reading the issues back, I hadn't noticed this used such a narrow macOS version. Perhaps we should gate the `PRAGMA` query under this platform check, to avoid an unnecessary query on every connection.

**Review by carltongibson — APPROVED** [review]
Good. 

This resolves the issue for a Python using an SQLite where `legacy_alter_table` defaults to `on` — specifically testing a `pyenv` installed Python dynamically linking the macOS provided SQLite on macOS 11.6.

I (currently) don't see this on macOS 12 because I'm using a Python from python.org, [which statically links SQLite, specifically to avoid issues with the macOS provided version](https://bugs.python.org/issue43773?@template=item#msg390551).

However, checking the provided SQLite on macOS 12, the same settings apply…

```
% sqlite3   
SQLite version 3.36.0 2021-06-18 18:58:49
...
sqlite> .dbconfig
...
 legacy_alter_table on
...
``` 

… presumably because this forms part of the expected base for Apple software building on it. Regardless, I didn't test but, the same error would come up with a Python linking against that. 

Thus, yep, let's have this. 

Thanks @adamchainz 👍

File: django/db/backends/sqlite3/features.py (hunk 2 above)
Comment (carltongibson) [reply django/db/backends/sqlite3/features.py:96]: Hey @adamchainz — sorry I missed this comment before merging. 

As per https://github.com/django/django/pull/15158#pullrequestreview-825140230 I think we need this always for SQLite on macOS — it's not version specific AFAICS. 

If you think there's a performance worry with the `PRAGMA` (really? 🤔) — then please do follow-up!

Comment (adamchainz) [reply django/db/backends/sqlite3/features.py:96]: > If you think there's a performance worry with the PRAGMA (really? 🤔) — then please do follow-up!

Yeah, it's not a huge worry, because it's SQLite. For client/server DB's we'd want to avoid the extra roundtrip.

**Comment by adamchainz** [comment]
That's a good point, I hadn't about it, but my pyenv Python is linked against brew's sqlite, not the system one. Thanks for checking.

**Comment by carltongibson** [comment]
> ...linked against brew's sqlite... 

Yes... — and then you're into Tom's point about `CFLAGS` and all the rest of it from the ticket. 

tl;dr this whole situation is more complicated that one'd wish. The change is a good fix — thanks for making it!
